Content.Server/Power/Generator/GeneratorSystem.cs

```diff
@@ -1,123 +1,258 @@
+                _popup.PopupEntity(Loc.GetString("generator-clogged", ("generator", uid)), uid, PopupType.SmallCaution);
+                SetFuelGeneratorOn(uid, false, gen);
+                continue;
+            }
+
+            supplier.Enabled = true;
 
             var upgradeMultiplier = _upgradeQuery.CompOrNull(uid)?.ActualScalar ?? 1f;
 
             supplier.MaxSupply = gen.TargetPower * upgradeMultiplier;
 
             var eff = 1 / CalcFuelEfficiency(gen.TargetPower, gen.OptimalPower, gen);
-
-            gen.RemainingFuel = MathF.Max(gen.RemainingFuel - (gen.OptimalBurnRate * frameTime * eff), 0.0f);
-            UpdateUi(uid, gen);
+            var consumption = gen.OptimalBurnRate * frameTime * eff;
+            RaiseLocalEvent(uid, new GeneratorUseFuel(consumption));
         }
     }
 
-    private void UpdateUi(EntityUid uid, FuelGeneratorComponent comp)
+    public float GetFuel(EntityUid generator)
     {
-        if (!_uiSystem.IsUiOpen(uid, GeneratorComponentUiKey.Key))
-            return;
+        GeneratorGetFuel getFuel = default;
+        RaiseLocalEvent(generator, ref getFuel);
+        return getFuel.Fuel;
+    }
 
-        _uiSystem.TrySetUiState(uid, GeneratorComponentUiKey.Key, new SolidFuelGeneratorComponentBuiState(comp));
+    public bool GetIsClogged(EntityUid generator)
+    {
+        GeneratorGetClogged getClogged = default;
+        RaiseLocalEvent(generator, ref getClogged);
+        return getClogged.Clogged;
     }
+
+    public void EmptyGenerator(EntityUid generator)
+    {
+        RaiseLocalEvent(generator, GeneratorEmpty.Instance);
+    }
+
+    private void UpdateState(EntityUid generator, FuelGeneratorComponent component)
+    {
+        _appearance.SetData(generator, GeneratorVisuals.Running, component.On);
+        _ambientSound.SetAmbience(generator, component.On);
+        if (!component.On)
+            Comp<PowerSupplierComponent>(generator).Enabled = false;
+    }
+}
+
+/// <summary>
+/// Raised by <see cref="GeneratorSystem"/> to calculate the amount of remaining fuel in the generator.
+/// </summary>
+[ByRefEvent]
+public struct GeneratorGetFuel
+{
+    public float Fuel;
+}
+
+/// <summary>
+/// Raised by <see cref="GeneratorSystem"/> to check if a generator is "clogged".
+/// For example there's bad chemicals in the fuel tank that prevent starting it.
+/// </summary>
+[ByRefEvent]
+public struct GeneratorGetClogged
+{
+    public bool Clogged;
+}
+
+/// <summary>
+/// Raised by <see cref="GeneratorSystem"/> to draw fuel from its adapters.
+/// </summary>
+/// <remarks>
+/// Implementations are expected to round fuel consumption up if the used fuel value is too small (e.g. reagent units).
+/// </remarks>
+public record struct GeneratorUseFuel(float FuelUsed);
+
+/// <summary>
+/// Raised by <see cref="GeneratorSystem"/> to empty a generator of its fuel contents.
+/// </summary>
+public sealed class GeneratorEmpty
+{
+    public static readonly GeneratorEmpty Instance = new();
 }
```

Comment:
append these with `Event` i.e. `GeneratorGetFuelEvent`. also can getfuel/getclogged just be record structs?

Reply:
Oh right I forgot record structs are mutable.